### ISSUE TITLE
[layout-flow] Add LightningCLI training and vendor step parity

### ISSUE THREAD

**Issue by creative-graphic-design-dev** [creative-graphic-design/design-generators#149]
## Summary

Add package-local LightningCLI training for LayoutFlow and verify step-level parity against the vendor Lightning/Hydra training loop.

Refs #148.

## Scope

First PR scope follows the train-ourselves unified spec:

- Add the `lib/traingen` workspace member for deterministic training parity helpers and LightningCLI support.
- Add `layout_flow.training` with `LayoutFlowLightningCLI`, `main`, `LayoutFlowTrainingModule`, `LayoutFlowDataModule`, `LayoutFlowH5Dataset`, `collate_layout_flow_batch`, and `LayoutFlowSeedMode`.
- Add LightningCLI YAML configs for RICO25/PubLayNet vendor-compatible training, strict parity training, and smoke runs.
- Add fast tests for config parsing, Hydra-key rejection, HDF5/collate behavior, random4 condition masking, scalar training-step behavior, and optimizer/scheduler wiring.
- Add S0-S2 parity hooks/scripts: static config/state comparison, fixed-batch pre-optimizer trace comparison, and one optimizer-step parameter comparison.

## Vendor Parity Requirements

The converted training path must reproduce the vendor `training_step` order:

```text
cond_mask -> x0/x1 -> t -> xt/ut -> vt -> flow_loss + 0.2 * geom_l1_loss
```

Required trace points include `bbox`, `type`, `mask`, `length`, stable sample id/HDF5 key when available, `cond_mask`, `x0`, `x1`, `t`, `xt`, `ut`, `vt`, `flow_loss`, `geom_l1_loss`, `train_loss`, gradients, optimizer state, post-step parameter hashes/allclose, and LR.

LayoutFlow-specific defaults:

- PubLayNet: `num_labels=6`, `sampler_out_dim=7`
- RICO/RICO25: `num_labels=26`, `sampler_out_dim=9`
- default condition policy: `random4`
- default loss: MSE plus `geom_l1_loss` with weight `0.2`
- optimizer: AdamW, `lr=0.0005`, `betas=[0.9, 0.98]`
- scheduler: ReduceLROnPlateau on `FID_Layout`

## Verification

Run S0-S2 parity on one explicitly selected GPU:

```bash
CUDA_VISIBLE_DEVICES=4 PARITY_REQUIRE=1 uv run --package layout-flow --extra training pytest models/layout-flow/tests/vendor_parity -m vendor_parity
```

The PR body must include measured S0-S2 parity results, commands, tolerances, and unresolved open questions from #148.

**Comment by creative-graphic-design-dev** [comment]
## S5 scheduler/FID validation deviation

S5 full-run comparison is being run with LR scheduling disabled and FID validation disabled on both implementations:

- package-local runs: `model.init_args.scheduler: null`, `model.init_args.fid_calc_every_n: 0`, `trainer.limit_val_batches: 0`
- vendor runs: `model.scheduler=null`, `model.fid_calc_every_n=0`, `+trainer.limit_val_batches=0`

This is an intentional S5 deviation from the original full-training config. The default scheduler is `ReduceLROnPlateau` and monitors `FID_Layout`; once FID validation is disabled, leaving the scheduler enabled causes a missing-metric failure and would also make the vendor/package-local LR trajectories incomparable if only one side disabled it. FID calculation also introduces validation-time sampling and independent RNG consumption, so S5 keeps it out of the training loop and compares the two training implementations under the same scheduler-free condition.

Scheduler-enabled equivalence remains covered by the staged S3/S4 validation surface rather than this S5 full-run comparison. The S5 report should carry this same deviation note.

Current sampled S5 status:

- ours RICO25: running, 315 complete epochs
- vendor RICO25: running, 314 complete epochs
- ours PubLayNet: running, 34 complete epochs
- vendor PubLayNet: running, 33 complete epochs

Logs and checkpoints are under `.cache/layout-flow/full-run/{vendor,ours}-{rico25,publaynet}/`; generated checkpoints/tensors are local only and are not committed.

**Comment by creative-graphic-design-dev** [comment]
Current LayoutFlow package-local training uses the original HDF5 files as the compatibility path so the Lightning data path can match the vendor training setup. A future follow-up should move package-local training data access to Hugging Face `datasets` and the creative-graphic-design org dataset mirrors where practical, while keeping the HDF5 path available for vendor-equivalence reruns.

**Comment by creative-graphic-design-dev** [comment]
## RICO25 S5 evaluation report

RICO25 S5 was evaluated without waiting for the PubLayNet jobs and without touching the PubLayNet processes. The completed RICO25 checkpoints were evaluated with the vendor `src/test.py` protocol on GPU 4 using eval seed 0.

### Checkpoints

| Path | Role | Epoch | Global step | Notes |
| --- | --- | ---: | ---: | --- |
| `.cache/layout-flow/full-run/vendor-rico25/checkpoints/last.ckpt` | vendor final | 1000 | 62000 | Used directly by vendor `test.py`. |
| `.cache/layout-flow/full-run/ours-rico25/checkpoints/last-v1.ckpt` | ours final | 1000 | 62000 | Converted to a vendor-protocol temporary checkpoint for `test.py`. |
| `.cache/layout-flow/full-run/ours-rico25/checkpoints/last.ckpt` | stale earlier run | 1 | 62 | Not used for S5. |

No separate best checkpoint was present in the RICO25 run directories; the completed `last` checkpoints above were used for final evaluation. The temporary vendor-protocol checkpoint for ours is `.cache/layout-flow/full-run/ours-rico25/eval-vendor-protocol/last-v1-vendor-compatible.ckpt`, with `model.backbone.*` keys normalized to `model.*`.

### Vendor `test.py` metrics

Protocol: `dataset=RICO`, `model=LayoutFlow`, `task=uncond`, `cond_mask=uncond`, `ode_solver=euler`, `model.inference_steps=100`, `calc_miou=True`, `multirun=False`, seed 0, `pretrained/fid_rico.pth.tar`, and the vendor RICO test tensor `pretrained/rico_test.pt`.

| Checkpoint | FID ↓ | Alignment ↓ | Overlap ↓ | mIoU ↑ |
| --- | ---: | ---: | ---: | ---: |
| vendor final | 6.5372 | 0.2084 | 0.6070 | 0.5755 |
| ours final | 5.0896 | 0.2259 | 0.5988 | 0.5161 |
| ours - vendor | -1.4476 | +0.0175 | -0.0082 | -0.0594 |

### Fixed-epoch training loss comparison

| Epoch | ours `train_loss_epoch` | vendor `train_loss_epoch` | Diff |
| ---: | ---: | ---: | ---: |
| 0 | 0.715035558 | 0.710425496 | +0.004610062 |
| 10 | 0.223951384 | 0.227125451 | -0.003174067 |
| 100 | 0.192493036 | 0.192266017 | +0.000227019 |
| 500 | 0.179825991 | 0.180727810 | -0.000901818 |
| 900 | 0.176882684 | 0.176147014 | +0.000735670 |
| 999 | 0.175238043 | 0.174988434 | +0.000249609 |

Across 1000 common epochs, `train_loss_epoch` mean absolute difference was `0.000905751`, RMSE was `0.001364406`, and correlation was `0.998998098`. For the last 100 epochs, mean relative train-loss difference was `0.001748`. Final train-loss relative difference was `0.001426`.

### Conversion and smoke

The ours final checkpoint was converted to `.cache/layout-flow/full-run/ours-rico25/converted-rico25` with:

```bash
uv run --package layout-flow python models/layout-flow/scripts/convert_original_checkpoint.py \
  --dataset rico25 \
  --checkpoint .cache/layout-flow/full-run/ours-rico25/checkpoints/last-v1.ckpt \
  --output-dir .cache/layout-flow/full-run/ours-rico25/converted-rico25
```

`from_pretrained` smoke passed:

```text
converted-rico25 torch.Size([1, 20, 4]) torch.Size([1, 20]) LayoutGenerationOutput
```

### S5 verdict

Training dynamics are statistically aligned for RICO25 (`PASS` for the loss-curve criterion used here: final train-loss relative difference `0.001426`, last-100 mean relative difference `0.001748`). Generated-layout metric equivalence is not fully established in the single-seed vendor protocol run: ours improves FID and Overlap, but Alignment is worse and mIoU drops by `0.0594`. Overall RICO25 S5 should be treated as `CHECK / partial pass` rather than a clean equivalence pass until an agreed tolerance or multiseed generated-metric comparison is defined.

Logs:

- `.cache/layout-flow/full-run/eval-rico25/vendor-protocol/vendor/eval.log`
- `.cache/layout-flow/full-run/eval-rico25/vendor-protocol/ours/eval.log`

**Comment by creative-graphic-design-dev** [comment]
## LayoutFlow RICO25 S5 n=3 final result

RICO25 multi-seed full training is complete for vendor and ours with seeds `42975`, `42976`, and `42977`. The final generated-layout metric comparison uses the vendor evaluation protocol and local checkpoints/assets only; no artifacts were committed.

### Per-seed metrics

| Run | FID | Alignment | Overlap | mIoU |
| --- | ---: | ---: | ---: | ---: |
| vendor-42975 | 6.4180 | 0.2071 | 0.6103 | 0.5845 |
| vendor-42976 | 5.5743 | 0.2242 | 0.5374 | 0.5449 |
| vendor-42977 | 7.1798 | 0.2396 | 0.5728 | 0.5598 |
| ours-42975 | 5.1551 | 0.2320 | 0.5899 | 0.5490 |
| ours-42976 | 5.4194 | 0.2299 | 0.5507 | 0.5679 |
| ours-42977 | 6.5588 | 0.2458 | 0.5784 | 0.5516 |

### Mean +/- std

| Group | n | FID | Alignment | Overlap | mIoU |
| --- | ---: | ---: | ---: | ---: | ---: |
| vendor | 3 | 6.3907 +/- 0.8031 | 0.2236 +/- 0.0163 | 0.5735 +/- 0.0365 | 0.5631 +/- 0.0200 |
| ours | 3 | 5.7111 +/- 0.7459 | 0.2359 +/- 0.0086 | 0.5730 +/- 0.0202 | 0.5562 +/- 0.0102 |

### Protocol and local evidence

- Metrics CSV: `.cache/layout-flow/full-run/eval-rico25-n3/vendor-protocol/metrics.csv`
- Summary CSV: `.cache/layout-flow/full-run/eval-rico25-n3/vendor-protocol/summary_mean_std.csv`
- Evaluation entrypoint/protocol: `vendor/layout-flow/src/test.py`, `model=LayoutFlow`, `dataset=RICO`, `task=uncond`, `cond_mask=uncond`, `ode_solver=euler`, `model.inference_steps=100`, `calc_miou=True`, `multirun=False`, fixed `LAYOUTFLOW_EVAL_SEED=42975`.
- Final ours-42977 eval used `CUDA_VISIBLE_DEVICES=5`.
- RICO25 data path override: `.cache/layout-flow/data/rico25`.
- Pretrained FID assets: `vendor/layout-flow/pretrained`.
- Checkpoints are local under `.cache`; no generated artifacts or checkpoints were committed.
- Figure paths for owner attachment:
  - `.cache/layout-flow/full-run/figures/rico25_train_loss_all_epochs.png`
  - `.cache/layout-flow/full-run/figures/rico25_train_loss_last100_epochs.png`
  - `.cache/layout-flow/full-run/figures/rico25_train_loss_relative_diff.png`

### Verdict

S5 generated-layout metrics are `PASS / acceptable equivalence` under the n=3 vendor protocol. The generated-layout distributions overlap: ours has better mean FID, slightly worse mean Alignment, near-identical mean Overlap, and mIoU remains within vendor std. This resolves the prior single-seed mIoU concern for RICO25 S5, with the caveat that loss curves and generated metrics are local `.cache` evidence.

GPU/process follow-up after RICO25: no RICO25 training/eval/watcher processes remain. GPU 4 is 0 MiB. GPU 5 still shows non-RICO stale/other compute-app entries in `nvidia-smi`, but there is no matching RICO25 process in `ps`.

**Comment by creative-graphic-design-dev** [comment]
## LayoutFlow PubLayNet S5 evaluation report

Verdict: **PASS / acceptable equivalence for training dynamics**.

The PubLayNet S5 check is complete for the available final full-training checkpoints. This is **evaluation-seed n=3**, not training-seed n=3: only one full-training checkpoint exists for each implementation, so the three runs reuse the same final checkpoints with evaluation seeds 42975, 42976, and 42977. A true training-seed n=3 verdict would require additional full PubLayNet training runs.

Checkpoints:

| Run | Checkpoint | Epoch | Global step |
| --- | --- | ---: | ---: |
| vendor | `.cache/layout-flow/full-run/vendor-publaynet/checkpoints/last.ckpt` | 1000 | 609000 |
| ours | `.cache/layout-flow/full-run/ours-publaynet/checkpoints/last.ckpt` | 1000 | 609000 |
| ours, vendor-protocol eval | `.cache/layout-flow/full-run/ours-publaynet/eval-vendor-protocol/last-vendor-compatible.ckpt` | 1000 | 609000 |

Evaluation protocol: vendor `vendor/layout-flow/src/test.py` through the local wrapper forcing `torch.load(weights_only=False)`; `model=LayoutFlow`, `dataset=PubLayNet`, `task=uncond`, `cond_mask=uncond`, `ode_solver=euler`, `model.inference_steps=100`, `calc_miou=True`, `multirun=False`; data path `.cache/layout-flow/data/publaynet`; FID assets from `vendor/layout-flow/pretrained`; final runs used GPU 1. The ours vendor-protocol checkpoint normalizes `model.backbone.*` to `model.*` and copies vendor hparams.

Per evaluation seed:

| Eval seed | Run | FID | Alignment | Overlap | mIoU |
| ---: | --- | ---: | ---: | ---: | ---: |
| 42975 | vendor | 16.7338 | 0.1126 | 0.0316 | 0.4315 |
| 42975 | ours | 12.5132 | 0.1167 | 0.0307 | 0.4165 |
| 42976 | vendor | 16.7218 | 0.1035 | 0.0305 | 0.4317 |
| 42976 | ours | 11.5772 | 0.1189 | 0.0290 | 0.4166 |
| 42977 | vendor | 16.3122 | 0.1056 | 0.0319 | 0.4239 |
| 42977 | ours | 11.6245 | 0.1125 | 0.0312 | 0.4122 |

Evaluation-seed n=3 summary:

| Run | n | FID mean +/- std | Alignment mean +/- std | Overlap mean +/- std | mIoU mean +/- std |
| --- | ---: | ---: | ---: | ---: | ---: |
| vendor | 3 | 16.5893 +/- 0.2400 | 0.1072 +/- 0.0048 | 0.0313 +/- 0.0007 | 0.4290 +/- 0.0044 |
| ours | 3 | 11.9050 +/- 0.5273 | 0.1160 +/- 0.0033 | 0.0303 +/- 0.0012 | 0.4151 +/- 0.0025 |
| ours - vendor | 3 | -4.6843 | +0.0088 | -0.0010 | -0.0139 |

Loss comparison over the completed 1000 common epochs:

| Metric | Value |
| --- | ---: |
| final vendor `train_loss_epoch` | 0.1530758440 |
| final ours `train_loss_epoch` | 0.1529803425 |
| final diff, ours - vendor | -0.0000955015 |
| mean absolute train-loss diff | 0.0002493398 |
| RMSE | 0.0003080925 |
| correlation | 0.9995266371 |
| last-100 mean relative diff | 0.0017203791 |

Interpretation: training dynamics match very closely. Generated-layout metrics are mixed but acceptable under this single-training-seed plus evaluation-seed n=3 check: ours improves FID and overlap slightly, while alignment is worse by about 0.0088 absolute and mIoU is lower by about 0.0139 absolute.

Evidence paths:

- `.cache/layout-flow/full-run/eval-publaynet/vendor-protocol/metrics.csv`
- `.cache/layout-flow/full-run/eval-publaynet/vendor-protocol-seed42976/metrics.csv`
- `.cache/layout-flow/full-run/eval-publaynet/vendor-protocol-seed42977/metrics.csv`
- `.cache/layout-flow/full-run/eval-publaynet/vendor-protocol-eval-seed-n3-metrics.csv`
- `.cache/layout-flow/full-run/eval-publaynet/vendor-protocol-eval-seed-n3-summary.csv`

**Comment by creative-graphic-design-dev** [comment]
## LayoutFlow PubLayNet S5 training-seed n=3 final result

PubLayNet multi-seed full training is complete for vendor and package-local runs with training seeds `42975`, `42976`, and `42977`. The final generated-layout metric comparison uses the vendor `test.py` protocol with fixed evaluation RNG, `model=LayoutFlow`, `dataset=PubLayNet`, `task=uncond`, `cond_mask=uncond`, `model.inference_steps=100`, and `calc_miou=True`. No checkpoints or generated artifacts were committed.

### Checkpoint completion

All six final checkpoints reached epoch `1000` and global step `609000`.

### Per-training-seed metrics

| Run | FID | Alignment | Overlap | mIoU |
| --- | ---: | ---: | ---: | ---: |
| vendor-42975 | 16.7338 | 0.1126 | 0.0316 | 0.4315 |
| vendor-42976 | 12.0102 | 0.1322 | 0.0559 | 0.4132 |
| vendor-42977 | 13.0819 | 0.1208 | 0.0294 | 0.4239 |
| ours-42975 | 12.5132 | 0.1167 | 0.0307 | 0.4165 |
| ours-42976 | 13.5760 | 0.1162 | 0.0440 | 0.4137 |
| ours-42977 | 14.8628 | 0.1152 | 0.0341 | 0.4150 |

### Mean +/- std

| Group | n | FID | Alignment | Overlap | mIoU |
| --- | ---: | ---: | ---: | ---: | ---: |
| vendor | 3 | 13.9420 +/- 2.4765 | 0.1219 +/- 0.0098 | 0.0390 +/- 0.0147 | 0.4229 +/- 0.0092 |
| ours | 3 | 13.6507 +/- 1.1766 | 0.1160 +/- 0.0008 | 0.0363 +/- 0.0069 | 0.4151 +/- 0.0014 |

### Verdict

S5 generated-layout metrics are `PASS / acceptable equivalence` under the training-seed n=3 vendor protocol. The package-local runs have comparable or better mean FID, Alignment, and Overlap, with mean mIoU lower by `0.0078` absolute and still within the vendor run-to-run standard deviation.
